girder/api/rest.py

```diff
@@ -59,18 +59,29 @@ def getUrlParts(url=None):
 def getApiUrl(url=None):
     """
     In a request thread, call this to get the path to the root of the REST API.
-    The returned path does *not* end in a forward slash.
+    The returned path does *not* end in a forward slash.  If no url is
+    specified, and there is a Referer header that includes the REST API root,
+    use that for determining the root.  This allows detection of the root from
+    within and outside of a proxy.
 
     :param url: URL from which to extract the base URL. If not specified, uses
-        `cherrypy.url()`
+        `cherrypy.url()` or the config value from ['server']['api_url'].
     """
+    apiPhrase = '/api/v1'
     url = url or cherrypy.url()
-    idx = url.find('/api/v1')
+    if url is None and 'Referer' in cherrypy.request.headers:
```

Comment:
Will this block ever be entered? Wouldn't `cherrypy.url()` have to return `None` for that?